File: LXUtilities/UIKitExtensions/UIScrollView+LXExtension.m
//
//  UIScrollView+LXExtension.m
//
//  Created by 从今以后 on 2017/4/20.
//  Copyright © 2017年 从今以后. All rights reserved.
//

#import "UIScrollView+LXExtension.h"

@implementation UIScrollView (LXExtension)

- (BOOL)lx_isContentAtTop
{
    if (@available(iOS 11, *)) {
        return self.contentOffset.y + self.adjustedContentInset.top <= 0;
    }
    return self.contentOffset.y + self.contentInset.top <= 0;
}

- (BOOL)lx_isContentAtBottom
{
    CGFloat contentInsetBottom;
    if (@available(iOS 11, *)) {
        contentInsetBottom = self.adjustedContentInset.bottom;
    } else {
        contentInsetBottom = self.contentInset.bottom;
    }

    if (self.contentOffset.y + self.bounds.size.height - contentInsetBottom >= self.contentSize.height) {
        return YES;
    }

    return NO;
}

- (void)lx_scrollToTopAnimated:(BOOL)animated
{
    CGPoint offset;
    if (@available(iOS 11.0, *)) {
        offset = CGPointMake(0, -self.adjustedContentInset.top);
    } else {
        offset = CGPointMake(0, -self.contentInset.top);
    }
    [self setContentOffset:offset animated:animated];
}

- (void)lx_scrollToBottomAnimated:(BOOL)animated
{
    CGPoint offset;
    if (@available(iOS 11.0, *)) {
        offset = CGPointMake(0, self.contentSize.height - self.bounds.size.height + self.adjustedContentInset.bottom);
    } else {
        offset = CGPointMake(0, self.contentSize.height - self.bounds.size.height + self.contentInset.bottom);
    }
    [self setContentOffset:offset animated:animated];
}

@end

File: LXUtilities/UIKitExtensions/Swift/UIScrollView+LXExtension.swift
//
//  UIScrollView+LXExtension.swift
//
//  Created by 从今以后 on 2017/10/23.
//  Copyright © 2017年 从今以后. All rights reserved.
//

import UIKit

extension Swifty where Base: UIScrollView {

    func isContentAtTop() -> Bool {
        if #available(iOS 11, *) {
            return base.contentOffset.y + base.adjustedContentInset.top <= 0
        }
        return base.contentOffset.y + base.contentInset.top <= 0
    }

    func isContentAtBottom() -> Bool {
        let contentInsetBottom: CGFloat = {
            if #available(iOS 11, *) {
                return base.adjustedContentInset.bottom
            }
            return base.contentInset.bottom
        }()

        if base.contentOffset.y + base.bounds.height - contentInsetBottom >= base.contentSize.height {
            return true
        }

        return false
    }

    func scrollToTop(animated: Bool) {
        guard !isContentAtTop() else {
            return
        }
        base.scrollRectToVisible(CGRect(x: 0, y: 0, width: 1, height: 1), animated: animated)
    }

    func scrollToBottom(animated: Bool) {
        guard !isContentAtBottom() else {
            return
        }
        let y = base.contentSize.height - 1
        base.scrollRectToVisible(CGRect(x: 0, y: y, width: 1, height: 1), animated: animated)
    }
}
